File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_21.cpp
// *****
// *   *
// *   *
// *   *
// *****

// BRUTE FORCE
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            if (i == 1 || i == n || j == 1 || j == n)
                cout << "*";
            else
                cout << " ";
        }
        cout << endl;
    }
    return 0;
}

// BETTER
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string border(n, '*');
    string middle = "*" + string(n - 2, ' ') + "*";

    cout << border << endl;
    for (int i = 0; i < n - 2; i++)
    {
        cout << middle << endl;
    }
    cout << border << endl;
    return 0;
}

// BEST
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string border(n, '*');
    string middle = "*" + string(n - 2, ' ') + "*";

    cout << border << '\n';
    cout << string(n - 2, middle + '\n');
    cout << border << '\n';
    return 0;
}

File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_21.py
# *****
# *   *
# *   *
# *   *
# *****

# BRUTE FORCE
n = 5
for i in range(1, n + 1):
    line = ""
    for j in range(1, n + 1):
        if i == 1 or i == n or j == 1 or j == n:
            line += "*"
        else:
            line += " "
    print(line)

# BETTER
n = 5
border = "*" * n
middle = "*" + " " * (n - 2) + "*"
print(border)
for _ in range(n - 2):
    print(middle)
print(border)

# BEST
n = 5
border = "*" * n
middle = "*" + " " * (n - 2) + "*"
print('\n'.join([border] + [middle] * (n - 2) + [border]))